File: apps/web/lib/api/ai.ts
import { API_URL } from "@/lib/api/client";
import type { AIInsightsFixture } from "@/lib/fixtures/types";

export type FaqChatMessage = {
  role: "user" | "assistant";
  content: string;
};

export type FaqChatResponse = {
  answer: string;
  sources: Array<{ sourceType?: string; sourceId?: string; excerpt: string }>;
  sessionId: string;
};

export type ReviewSummaryResponse = {
  productId: string;
  summary: string;
  generatedAt: string;
};

export type AgentAction = {
  id: string;
  label: string;
};

export type AdminAgentResponse = {
  agentType: string;
  action: string | null;
  result: string;
  toolsCalled?: unknown[];
  logId: string;
};

export class AiApiError extends Error {
  constructor(
    message: string,
    public status: number,
  ) {
    super(message);
    this.name = "AiApiError";
  }
}

export async function fetchReviewInsights(): Promise<AIInsightsFixture> {
  const res = await fetch(`${API_URL}/reviews/insights`, {
    headers: { Accept: "application/json" },
    next: { revalidate: 300 },
  });
  if (!res.ok) {
    throw new AiApiError("Failed to load review insights", res.status);
  }
  return res.json() as Promise<AIInsightsFixture>;
}

export async function fetchProductReviewSummary(
  slug: string,
): Promise<ReviewSummaryResponse | null> {
  const res = await fetch(`${API_URL}/products/${encodeURIComponent(slug)}/review-summary`, {
    headers: { Accept: "application/json" },
    next: { revalidate: 300 },
  });
  if (res.status === 404) {
    return null;
  }
  if (!res.ok) {
    throw new AiApiError("Failed to load review summary", res.status);
  }
  return res.json() as Promise<ReviewSummaryResponse>;
}

export async function sendFaqChatMessage(
  message: string,
  sessionId?: string,
): Promise<FaqChatResponse> {
  const res = await fetch(`${API_URL}/chat/faq`, {
    method: "POST",
    headers: { "Content-Type": "application/json", Accept: "application/json" },
    body: JSON.stringify({ message, sessionId }),
  });
  if (!res.ok) {
    const body = await res.text();
    throw new AiApiError(body || "Chat unavailable", res.status);
  }
  return res.json() as Promise<FaqChatResponse>;
}

export async function fetchAgentActions(): Promise<Record<string, AgentAction[]>> {
  const { adminFetch } = await import("@/lib/api/admin");
  const data = await adminFetch<{ agents: Record<string, AgentAction[]> }>("/admin/ai/actions");
  return data.agents;
}

export async function runAdminAgentAction(
  agentType: "product" | "inventory" | "marketing" | "reviews",
  action: string,
): Promise<AdminAgentResponse> {
  const { adminFetch } = await import("@/lib/api/admin");
  return adminFetch<AdminAgentResponse>(`/admin/ai/agents/${agentType}`, {
    method: "POST",
    body: JSON.stringify({ action }),
  });
}

/** Legacy panel helper — maps free-text prompt to agent action endpoint */
export async function runAdminAgent(
  agentType: "product" | "inventory" | "marketing" | "reviews" | "image",
  prompt: string,
): Promise<AdminAgentResponse> {
  if (agentType === "image") {
    return {
      agentType,
      action: null,
      result: "Image prep agent will be wired in Phase 7.",
      logId: "stub",
    };
  }
  return runAdminAgentAction(agentType, prompt);
}

export async function triggerKnowledgeReindex(): Promise<{ taskId?: string; message: string }> {
  const { adminFetch } = await import("@/lib/api/admin");
  return adminFetch<{ taskId?: string; message: string }>("/admin/ai/reindex", {
    method: "POST",
  });
}

export async function regenerateProductReviewSummary(
  productId: string,
): Promise<{ taskId?: string; message: string }> {
  const { adminFetch } = await import("@/lib/api/admin");
  return adminFetch<{ taskId?: string; message: string }>(
    `/admin/products/${productId}/reviews/summarize/regenerate`,
    { method: "POST" },
  );
}

export async function summarizeProductReviews(
  productId: string,
): Promise<ReviewSummaryResponse> {
  const { adminFetch } = await import("@/lib/api/admin");
  return adminFetch<ReviewSummaryResponse>(
    `/admin/products/${productId}/reviews/summarize`,
    { method: "POST" },
  );
}

File: apps/api/app/schemas/ai.py
from __future__ import annotations

from pydantic import BaseModel, Field


class FaqChatRequestSchema(BaseModel):
    message: str = Field(..., min_length=1, max_length=2000)
    session_id: str | None = Field(None, alias="sessionId")

    model_config = {"populate_by_name": True}


class FaqChatSourceSchema(BaseModel):
    source_type: str | None = Field(None, alias="sourceType")
    source_id: str | None = Field(None, alias="sourceId")
    excerpt: str

    model_config = {"populate_by_name": True}


class FaqChatResponseSchema(BaseModel):
    answer: str
    sources: list[FaqChatSourceSchema]
    session_id: str = Field(..., alias="sessionId")

    model_config = {"populate_by_name": True}


class AIInsightsSchema(BaseModel):
    status: str
    summary: str
    positive_themes: list[str] = Field(..., alias="positiveThemes")
    concerns: list[str]
    trending_praise: list[str] = Field(..., alias="trendingPraise")
    last_updated: str | None = Field(None, alias="lastUpdated")

    model_config = {"populate_by_name": True}


class ReviewSummarySchema(BaseModel):
    product_id: str = Field(..., alias="productId")
    summary: str
    generated_at: str = Field(..., alias="generatedAt")

    model_config = {"populate_by_name": True}


class AdminAgentRequestSchema(BaseModel):
    action: str | None = Field(None, min_length=1, max_length=128)
    prompt: str | None = Field(None, min_length=1, max_length=4000)
    context: dict | None = None


class AdminAgentActionSchema(BaseModel):
    id: str
    label: str


class AdminAgentActionsSchema(BaseModel):
    agents: dict[str, list[AdminAgentActionSchema]]


class AdminAgentResponseSchema(BaseModel):
    agent_type: str = Field(..., alias="agentType")
    action: str | None = None
    result: str
    tools_called: list | None = Field(None, alias="toolsCalled")
    log_id: str = Field(..., alias="logId")

    model_config = {"populate_by_name": True}


class ReindexResponseSchema(BaseModel):
    status: str
    task_id: str | None = Field(None, alias="taskId")
    message: str

    model_config = {"populate_by_name": True}
